use-cases/src/Contracts/CrowdFunding.hs

```diff
@@ -0,0 +1,119 @@
+-- in a single transaction. This is, of course, limited by the maximum
+-- number of inputs a transaction can have.
+
+module Contracts.CrowdFunding (contribute, collect, refund)
+where
+
+import           Control.Monad (void)
+import           Plutus
+{-# ANN module "HLint: ignore Avoid restricted function" #-}
+
+
+-- Functionality for campaign contributors
+-- --
+
+-- Contribute funds to the campaign (contributor)
+--
+contribute :: PubKey -> Value -> T TxOutRef
+contribute ownerPubKey value
+  = do
+    { assert (value > 0)
+    ; contributorPubKey <- lookupMyPubKey
+    ; myPayment         <- createPayment (value + standardTxFee)
+    ; submitTransaction Tx
+      { txInputs  = [myPayment]
+      , txOutputs = [TxOutScript
+                       value
+                       (contributionScript ownerPubKey)
+                       [| contributorPubKey |] -- data script: ought to be lifted into PLC at coordination runtime
+                    ]
+      }
+    }
+    -- the transaction above really ought to be merely a transaction *template* and the transaction fee ought to be
+    -- added by the Wallet API Plutus library on the basis of the size and other costs of the transaction
+
+contributionScript :: PubKey -> Validator
+contributionScript ownerPubKey
+  = -- needs a splice adding the function triggering the Plutus Tx compiler plugin (compiled at coordination compile time)
+    --
+    -- What data is available inside the validator script? 
+    -- 
+    -- The transaction references its input via their hashes, so we need to 
+    -- de-reference them to be able to look at the coin values of their
+    -- outpus. De-referencing of a hash should probably also cost money 
+    -- (otherwise we could jam the validator by asking for data all the way
+    -- back to the beginning of the chain)
+    --
+    -- (see https://github.com/input-output-hk/cardano-sl/blob/develop/docs/block-processing/types.md for a description of the data types that exist in a block)
+    -- 
+    [| \redeemer txn contributorPubKey ->
+      let inputFunds = sum $ (sum $ view outputs) <$> inputs txn
+          payToOwner = currentBlockHeight > paymentDeadline &&
+                       currentBlockHeight < collectionDeadline &&
+                       inputFunds > fundingGoal &&
+                       checkSig redeemer ownerPubKey
+          refund     = currentBlockHeight > collectionDeadline &&
+                       checkSig redeemer contributorPubKey
+      in
+        payToOwner || refund
+    |]
+
+-- We can turn a validation script (belonging to an address) into an event
+-- trigger (specific to a wallet) by partially evaluating the script.
+--
+-- For example, as a contributor, the `payToOwner` branch of the script above
+-- is never true because we don't have the owner's private key and
+-- `checkSig redeemer ownerPubKey` will always be false. On the other hand,
+-- `checkSig redeemer contributorPubKey` can be made true (because we know the

```

Comment:
This seems to be brushing some complexity under the rug - we need to know what value we're going to submit as the redeemer. Which we might know in advance, or we might not (perhaps it's controlled by the wallet user).

Reply:
I do like the idea of taking the script and partially evaluating it into a trigger, but I'm not sure it'll obviously work in general.

Reply:
> we need to know what value we're going to submit as the redeemer

But whatever we submit as the redeemer script, we can only sign it with `ownerPubKey` if we have the private key. So this branch is going to be false for any transaction that is generated by the contributor. (assuming they are not also the owner)

Reply:
I believe that you as a human can reason about that, I am less confident that we can do it automatically ;)